cdc/puller/unified_sorter.go

```diff
@@ -0,0 +1,915 @@
+	}
+	return nil
+}
+
+type memorySorterBackEnd struct {
+	events    []*model.PolymorphicEvent
+	readIndex int
+}
+
+func (m *memorySorterBackEnd) readNext() (*model.PolymorphicEvent, error) {
+	if m.readIndex >= len(m.events) {
+		return nil, nil
+	}
+	ret := m.events[m.readIndex]
+	m.readIndex += 1
+	return ret, nil
+}
+
+func (m *memorySorterBackEnd) writeNext(event *model.PolymorphicEvent) error {
+	m.events = append(m.events, event)
+	return nil
+}
+
+func (m *memorySorterBackEnd) getSize() int {
+	return -1
+}
+
+func (m *memorySorterBackEnd) flush() error {
+	return nil
+}
+
+func (m *memorySorterBackEnd) reset() error {
+	m.events = m.events[0:0]
+	m.readIndex = 0
+	return nil
+}
+
+type backEndPool struct {
+	memoryUseEstimate int64
+	fileNameCounter   uint64
+	memPressure       int32
+	cache             [256]unsafe.Pointer
+	dir               string
+}
+
+func newBackEndPool(dir string) *backEndPool {
+	ret := &backEndPool{
+		memoryUseEstimate: 0,
+		fileNameCounter:   0,
+		dir:               dir,
+	}
+
+	go func() {
+		ticker := time.NewTicker(5 * time.Second)
+
+		for {
+			<-ticker.C
+
+			// update memPressure
+			m, err := memory.Get()
+			if err != nil {
+				log.Fatal("unified sorter: getting system memory usage failed", zap.Error(err))
+			}
+
+			memPressure := m.Used * 100 / m.Total
+			atomic.StoreInt32(&ret.memPressure, int32(memPressure))
+			if memPressure > 50 {
+				log.Debug("unified sorter: high memory pressure", zap.Uint64("memPressure", memPressure))
+				// Increase GC frequency to avoid necessary OOM
+				debug.SetGCPercent(10)
+			} else {
+				debug.SetGCPercent(100)
+			}
+
+			// garbage collect temporary files in batches
+			freedCount := 0
+			for i := range ret.cache {
+				ptr := &ret.cache[i]
+				innerPtr := atomic.SwapPointer(ptr, nil)
+				if innerPtr == nil {
+					continue
+				}
+				backEnd := (*fileSorterBackEnd)(innerPtr)
+				err := backEnd.free()
+				if err != nil {
+					log.Warn("Cannot remove temporary file for sorting", zap.String("file", backEnd.name), zap.Error(err))
+				} else {
+					log.Info("Temporary file removed", zap.String("file", backEnd.name))
+					freedCount += 1
+				}
+				if freedCount >= 16 {
+					freedCount = 0
+					break
+				}
+			}
+		}
+	}()
+
+	return ret
+}
+
+func (p *backEndPool) alloc(ctx context.Context) (sorterBackEnd, error) {
+	sorterConfig := config.GetSorterConfig()
+	if atomic.LoadInt64(&p.memoryUseEstimate) < int64(sorterConfig.MaxMemoryConsumption) &&
+		atomic.LoadInt32(&p.memPressure) < int32(sorterConfig.MaxMemoryPressure) {
+
+		ret := new(memorySorterBackEnd)
+		atomic.AddInt64(&p.memoryUseEstimate, int64(sorterConfig.ChunkSizeLimit))
+		return ret, nil
+	}
+
+	for i := range p.cache {
+		ptr := &p.cache[i]
+		ret := atomic.SwapPointer(ptr, nil)
+		if ret != nil {
+			return (*fileSorterBackEnd)(ret), nil
+		}
+	}
+
+	fname := fmt.Sprintf("%s/sort-%d-%d", p.dir, os.Getpid(), atomic.AddUint64(&p.fileNameCounter, 1))
```

Comment:
Supposing two processors in the same capture to replicate two changefeeds,  they will share the same filenames, whether there exists conflict?